core/amber/src/main/python/core/util/customized_queue/linked_blocking_multi_queue.py

```diff
@@ -0,0 +1,436 @@
+            self.count: AtomicInteger = AtomicInteger()
+            self.enabled: bool = True
+            self.head: LinkedBlockingMultiQueue.Node = LinkedBlockingMultiQueue.Node(
+                None
+            )
+            self.last: Optional[LinkedBlockingMultiQueue.Node[T]] = self.head
+
+        def clear(self) -> None:
+            self.fully_lock()
+            try:
+                h: LinkedBlockingMultiQueue.Node[T] = self.head
+                p: LinkedBlockingMultiQueue.Node = h.next
+                while p is not None:
+                    h.next = h
+                    p.item = None
+                    h = p
+                    p = h.next
+                self.head = self.last
+                old_count = self.count.get_and_set(0)
+                if self.enabled:
+                    self.owner.total_count.get_and_dec(old_count)
+            finally:
+                self.fully_unlock()
+
+        def disable(self) -> None:
+            self.fully_lock()
+            try:
+                if not self.enabled:
+                    return
+                self.owner.total_count.dec(self.count.value)
+                self.enabled = False
+            finally:
+                self.fully_unlock()
+
+        def enable(self) -> None:
+            self.fully_lock()
+            try:
+                if self.enabled:
+                    return
+                self.enabled = True
+
+                # potentially unlock waiting polls
+                c = self.count.value
+                if c > 0:
+                    self.owner.total_count.inc(c)
+                    self.owner.not_empty.notify()
+
+            finally:
+                self.fully_unlock()
+
+        def is_enabled(self) -> bool:
+            self.owner.take_lock.acquire()
+            try:
+                return self.enabled
+            finally:
+                self.owner.take_lock.release()
+
+        def enqueue(self, node: LinkedBlockingMultiQueue.Node[T]) -> None:
+            self.last.next = node
+            self.last = node
+
+        def __str__(self) -> str:
+            res = ""
+            h = self.head
+            while h.next is not None:
+                res += h.next.item
+                res += " -> "
+                h = h.next
+            return res
+
+        def size(self) -> int:
+            return self.count.value
+
+        def is_empty(self) -> bool:
+            return self.size() == 0
+
+        def put(self, obj: T) -> None:
+            old_size = -1
+            node = LinkedBlockingMultiQueue.Node(obj)
+            self.put_lock.acquire()
+            try:
+                self.enqueue(node)
+                self.count.inc()
+                if self.enabled:
+                    old_size = self.owner.total_count.get_and_inc()
+            finally:
+                self.put_lock.release()
+
+            if old_size == 0:
+                self.owner._signal_not_empty()
+
+        def remove(self, obj: T) -> bool:
```

Comment:
we are also not checking for null here, if the `obj` passed in is `None`, I think we might wrongly remove the special head node?

Reply:
Changed to return False when removing a None value. As discussed above, it might be easy to drop the support of NoneType as a queue element for now.